### PR TITLE
persist: add metrics to s3 timeouts; timeouts can be dynamically set

### PR DESCRIPTION
**Pull request by pH14**
Quick follow-up to https://github.com/MaterializeInc/materialize/pull/17239 -- gives us metric coverage of our S3 client's internal timeouts, and sets up the values to be dynamically configurable (once we finish integrating LD + PersistConfig). Uses a dirty hack to hook into the S3 client's `sleep` impl to figure out which timeouts are requested, allowing us to substitute in a dynamic value and record when a timeout is hit.

### Motivation

Will make it easier to [drive down connection timeouts](https://github.com/MaterializeInc/materialize/issues/17227) + observe network outages.

I also think this makes a meaningful dent in the `Put timeouts on all s3 and crdb usage (would have caught networking outage)` M3 persist item

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a
  companion cloud PR to account for those changes that is tagged with
  the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

File: src/persist/src/cfg.rs [hunk 1]
```diff
@@ -36,6 +36,18 @@ pub enum BlobConfig {
     Mem,
 }
 
+/// Configuration knobs for [Blob].
+pub trait BlobKnobs: std::fmt::Debug + Send + Sync {
+    /// maximum time allowed for a network call, including retry attempts
```
Comment (danhhz) [inline src/persist/src/cfg.rs:41]: nit: capital letter at the beginning of doc comments please

File: src/persist/src/s3.rs [hunk 2]
```diff
@@ -46,6 +52,56 @@ pub struct S3BlobConfig {
...
+            ),
+            CONNECT_TIMEOUT_MARKER => (
+                self.knobs.connect_timeout(),
+                Some(self.metrics.connect_timeouts.clone()),
+            ),
+            READ_TIMEOUT_MARKER => (
+                self.knobs.read_timeout(),
+                Some(self.metrics.read_timeouts.clone()),
+            ),
+            duration => (duration, None),
```
Comment (danhhz) [inline src/persist/src/s3.rs:89]: when might we hit this case? is unexpected and something we'd want to detect?

**Review by danhhz — APPROVED** [review]
clever!

File: src/persist/src/s3.rs (hunk 2 above)
Comment (pH14) [reply src/persist/src/s3.rs:89]: the sleep impl is passed around to some deep internals in the SDK, and I believe it does get used internally for timeouts unrelated to the top-level client knobs

Comment (danhhz) [reply src/persist/src/s3.rs:88]: 👍

**Pull request merged**